lib/pos_block_producer.go

```diff
@@ -0,0 +1,145 @@
+}
+
+func NewPosBlockProducer(mm *MempoolManager, params *DeSoParams) *PosBlockProducer {
+	return &PosBlockProducer{
+		mm:     mm,
+		params: params,
+	}
+}
+
+// CreateBlockTemplate constructs a block with Fee-Time ordered transactions.
+// TODO: how do we set the following Header fields?
+//   - ProposerPublicKey
+//   - ProposerVotingPublicKey
+//   - ProposerRandomSeedHash
+//   - ProposedInView
+//   - ValidatorsVoteQC
+//   - ValidatorsTimeoutAggregateQC
+//   - ProposerVotePartialSignature
+//     Perhaps the caller of CreateBlockTemplate (server/consensus) will fill these out. The block is also unsigned.
+func (pbp *PosBlockProducer) CreateBlockTemplate(chainTip *BlockNode, latestBlockView *UtxoView) (*MsgDeSoBlock, error) {
+	// Fill out some initial header info
+	block := NewMessage(MsgTypeBlock).(*MsgDeSoBlock)
+	block.Header.Version = HeaderVersion2
+	block.Header.Height = uint64(chainTip.Height + 1)
+	block.Header.PrevBlockHash = chainTip.Hash
+	// TODO: determine block timestamp, blockTstamp
+	//  block.Header.SetTstampSecs(uint64(blockTstamp))
+
+	// Create a placeholder block reward transaction. We will update the amount later.
+	blockRewardOutput := &DeSoOutput{}
+	blockRewardOutput.AmountNanos = math.MaxUint64
+	blockRewardTxn := NewMessage(MsgTypeTxn).(*MsgDeSoTxn)
+	blockRewardTxn.TxOutputs = append(blockRewardTxn.TxOutputs, blockRewardOutput)
+	blockRewardTxn.TxnMeta = &BlockRewardMetadataa{
+		ExtraData: UintToBuf(0),
+	}
+	blockRewardTxnSizeBytes, err := blockRewardTxn.ToBytes(true)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error computing block reward txn size: ")
+	}
+
+	block.Txns = append(block.Txns, blockRewardTxn)
+
+	// TODO: we should probably have another parameter for this
+	maxBlockContentsSizeBytes := pbp.params.MinerMaxBlockSizeBytes
+
+	// 2. Get Fee-Time ordered transactions from the mempool and determine the pass/fail flags for each txn.
+	// TODO: We should also include per-transaction timestamps from the proposer.
+	feeTimeTxns, txnConnectStatusByIndex, maxUtilityFee, err := pbp.getBlockTransactions(chainTip, latestBlockView,
+		maxBlockContentsSizeBytes-uint64(len(blockRewardTxnSizeBytes)))
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error getting block transactions: ")
+	}
+	// Append the feeTImeTxns to our block.
+	block.Txns = append(block.Txns, feeTimeTxns...)
+	// Set the txnConnectStatusByIndex in the block and header.
+	// TODO: Add fixed-length bitset encoding.
+	block.TxnConnectStatusByIndex = txnConnectStatusByIndex
+	block.Header.TxnConnectStatusByIndexHash = HashBitset(txnConnectStatusByIndex)
+
+	// Update the block reward transaction with the correct amount.
+	blockRewardOutput.AmountNanos = maxUtilityFee
+
+	// Compute the merkle root for the block now that all of the transactions have been added.
+	merkleRoot, _, err := ComputeMerkleRoot(block.Txns)
+	if err != nil {
+		return nil, err
+	}
+	block.Header.TransactionMerkleRoot = merkleRoot
+
+	return block, nil
+}
+
```

Comment:
Nina is working on exposing the leader scheduled and block proposer's public keys from blockchain.go. You won't have her code available in time to pass the data needed into this function, but you could implement this function yourself so that it's ready for when her part of the code is done. This function is also also unit-testable so we can ensure it's bullet-proof.

Up to you. Any one of the three of us can do this part.

```suggestion
func SignPoSBlockTemplate(block *MsgDeSoBlock, proposer *ValidatorEntry, proposerPrivKey *bls.PrivateKey) (*MsgDeSoBlock, error) {
   // TODO lazynina or Aoensw4n or tholonious
}
```